File: Tree/DFS/dfs.java
package questions.Trees.BinaryTree.DFS;

import java.util.Scanner;

class TreeNode{
    int data;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) {
        data = x;
        left = right = null;
    }
}
public class dfs {
    TreeNode root;
    void insert(int key) {
        root = insertRec(root, key);
    }
    TreeNode insertRec(TreeNode root, int data)
    {
        Scanner sc = new Scanner(System.in);
        if (root == null) {
            root = new TreeNode(data);
            return root;
        }
        System.out.println("Enter choice : \n 1. Left \n 2. Right");
        int x = sc.nextInt();
        if (x==1)
            root.left = insertRec(root.left, data);
        else if (x==2)
            root.right = insertRec(root.right, data);
        return root;
    }
    public static void preOrder(TreeNode node)
    {
        if(node == null) return;

        System.out.print(node.data+" ");
        preOrder(node.left);
        preOrder(node.right);
    }
    public static void inOrder(TreeNode node)
    {
        if(node == null) return;

        inOrder(node.left);
        System.out.print(node.data+" ");
        inOrder(node.right);
    }
    public static void postOrder(TreeNode node)
    {
        if(node == null) return;

        postOrder(node.left);
        postOrder(node.right);
        System.out.print(node.data+" ");
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        dfs obj = new dfs();
        while(true){
            System.out.print("Enter Value : ");
            int x = sc.nextInt();
            if(x <0) break;
            obj.insert(x);
        }
        System.out.print("\nPreOrder -> ");
        preOrder(obj.root);
        System.out.print("\nInOrder -> ");
        inOrder(obj.root);
        System.out.print("\nPostOrder -> ");
        postOrder(obj.root);
    }
}

File: Tree/DFS/dfs.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def preorder(root):
    return [root.val] + preorder(root.left) + preorder(root.right) if root else []


def inorder(root):
    return inorder(root.left) + [root.val] + inorder(root.right) if root else []


def postorder(root):
    return postorder(root.left) + postorder(root.right) + [root.val] if root else []


root = TreeNode(10)
root.left = TreeNode(20)
root.right = TreeNode(30)
root.left.left = TreeNode(40)
root.left.right = TreeNode(50)
root.right.left = TreeNode(60)
root.right.right = TreeNode(70)

print(preorder(root))
print(inorder(root))
print(postorder(root))
